### PR TITLE
Output heatfluxes (total and maximum) per surface

### LINKED ISSUE #1531: Enable per-surface heatflux monitoring in history
**Is your feature request related to a problem? Please describe.**
In my application the heat flux over various surfaces is an important figure of merit of a specific design. Being able to monitor the average (or total) heat flux over different surfaces during runtime would be a good metric for judging convergence and quality of the design. Currently (as of 7.3.0)  this does not seem to be possible.

*A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]*
In my mesh I have defined various markers which indicate the surfaces of interest, where I would like the average heat-flux to evaluate over time. As suggested by the DryRun info I've indicated the surfaces with the MARKER_MONITORING specifier in the config file:

`MARKER_MONITORING = (NS_WALL_RESONATORTIP, NS_WALL_RESONATOR3, NS_WALL_RESONATOR2)`

However, the history and screen output shows only the heat flux aggregated over all surfaces.

**Describe the solution you'd like**
*A clear and concise description of what you want to happen.*

I would like the values of the HEAT group to behave just like those of the FLOW_COEFF group, where the heat flux is indicated on a per-surface basis. This would allow me to track the development of the heat fluxes with good time resolution, without having to write an excessive amount of surface files.

**Describe alternatives you've considered**
*A clear and concise description of any alternative solutions or features you've considered.*
Regularly exporting the entire surface data to discrete files and analyzing them during post-processing is a possible work-around, but leads to an excessive amount of files.

**Additional context**
*Add any other context or screenshots about the feature request here.*
My cases are highly transient and require quite some timesteps until a quasi-stationary limit cycle is reached. In the limit cycle the cycle-averaged net heat fluxes should be nearly zero. However, the magnitude of the heat transport is also of interest, which is why I would like to monitor these properties over time.

### PR DESCRIPTION
**Pull request by pcarruscag**
Closes #1531

### CODE REVIEW

File: SU2_CFD/src/output/CFlowOutput.cpp [hunk 1]
```diff
@@ -1183,14 +1183,58 @@ void CFlowOutput::SetAerodynamicCoefficients(CConfig *config, CSolver *flow_solv
...
+  AddHistoryOutput("MAXIMUM_HEATFLUX", "maxHF", ScreenOutputFormat::SCIENTIFIC, "HEAT", "Maximum heatflux across all surfaces set with MARKER_MONITORING.", HistoryFieldType::COEFFICIENT);
+
+  vector<string> Marker_Monitoring;
+  for (auto iMarker = 0u; iMarker < config->GetnMarker_Monitoring(); iMarker++) {
+    Marker_Monitoring.push_back(config->GetMarker_Monitoring_TagBound(iMarker));
+  }
+  /// DESCRIPTION:  Total heatflux
+  AddHistoryOutputPerSurface("TOTAL_HEATFLUX_ON_SURFACE", "HF", ScreenOutputFormat::SCIENTIFIC, "HEAT_SURF", Marker_Monitoring, HistoryFieldType::COEFFICIENT);
+  /// DESCRIPTION:  Total heatflux
+  AddHistoryOutputPerSurface("MAXIMUM_HEATFLUX_ON_SURFACE", "maxHF", ScreenOutputFormat::SCIENTIFIC, "HEAT_SURF", Marker_Monitoring, HistoryFieldType::COEFFICIENT);
```
Comment (TobiKattmann) [inline SU2_CFD/src/output/CFlowOutput.cpp:1203]: Why not follow the convention introduced with FLOW_COEFF_SURF and AERO_COEFF_SURF of not touching the names? If an `_ON_SURFACE` addition is made I would be in favor to be consistent and make it everywhere
```suggestion
  AddHistoryOutputPerSurface("TOTAL_HEATFLUX", "HF", ScreenOutputFormat::SCIENTIFIC, "HEAT_SURF", Marker_Monitoring, HistoryFieldType::COEFFICIENT);
  /// DESCRIPTION:  Total heatflux
  AddHistoryOutputPerSurface("MAXIMUM_HEATFLUX", "maxHF", ScreenOutputFormat::SCIENTIFIC, "HEAT_SURF", Marker_Monitoring, HistoryFieldType::COEFFICIENT);
```

File: SU2_CFD/src/output/CFlowOutput.cpp [hunk 2]
```diff
@@ -1183,14 +1183,58 @@ void CFlowOutput::SetAerodynamicCoefficients(CConfig *config, CSolver *flow_solv
...
+void CFlowOutput::SetHeatCoefficients(const CConfig* config, const CSolver* flow_solver) {
+
+  if (!config->GetViscous()) return;
+
+  SetHistoryOutputValue("TOTAL_HEATFLUX", flow_solver->GetTotal_HeatFlux());
+  SetHistoryOutputValue("MAXIMUM_HEATFLUX", flow_solver->GetTotal_MaxHeatFlux());
+
+  for (auto iMarker = 0u; iMarker < config->GetnMarker_Monitoring(); iMarker++) {
+    SetHistoryOutputPerSurfaceValue("TOTAL_HEATFLUX_ON_SURFACE", flow_solver->GetSurface_HF_Visc(iMarker), iMarker);
+    SetHistoryOutputPerSurfaceValue("MAXIMUM_HEATFLUX_ON_SURFACE", flow_solver->GetSurface_MaxHF_Visc(iMarker), iMarker);
```
Comment (TobiKattmann) [inline SU2_CFD/src/output/CFlowOutput.cpp:1216]: See comment above
```suggestion
    SetHistoryOutputPerSurfaceValue("TOTAL_HEATFLUX", flow_solver->GetSurface_HF_Visc(iMarker), iMarker);
    SetHistoryOutputPerSurfaceValue("MAXIMUM_HEATFLUX", flow_solver->GetSurface_MaxHF_Visc(iMarker), iMarker);
```

File: SU2_CFD/src/output/CMultizoneOutput.cpp [hunk 3]
```diff
@@ -181,15 +172,11 @@ void CMultizoneOutput::SetMultizoneHistoryOutputFields(const COutput* const* out
...
       /*--- Determine whether Maker_Analyze/Monitoring has to be used. ---*/
-      vector<string>* Marker;
+      auto* Marker = &Marker_Monitoring;
       if (group == "FLOW_COEFF_SURF")
         Marker = &Marker_Analyze;
-      else if (group == "AERO_COEFF_SURF")
-        Marker = &Marker_Monitoring;
-      else {
-        Marker = &Marker_Analyze;  // dummy to suppress maybe-uninitialized warning
+      else if (group != "AERO_COEFF_SURF" && group != "HEAT_SURF")
```
Comment (TobiKattmann) [inline SU2_CFD/src/output/CMultizoneOutput.cpp:178]: :+1:

File: SU2_CFD/src/output/CMultizoneOutput.cpp [hunk 4]
```diff
@@ -103,20 +103,11 @@ void CMultizoneOutput::LoadMultizoneHistoryData(const COutput* const* output, co
...
-        SU2_MPI::Error("Per Surface output group unknown: " + group, CURRENT_FUNCTION);
-
-      for (unsigned short iMarker = 0; iMarker < nMarker; iMarker++) {
-        SetHistoryOutputPerSurfaceValue(nameMultizone, item.second[iMarker].value, iMarker);
-      }// for iMarker
-    }// for HistPerSurfFields
+      unsigned short iMarker = 0;
+      for (const auto& field : item.second) {
+        SetHistoryOutputPerSurfaceValue(nameMultizone, field.value, iMarker++);
+      }
```
Comment (TobiKattmann) [inline SU2_CFD/src/output/CMultizoneOutput.cpp:109]: Nice, I knew there was a better way but my smooth brain wouldn't allow me finding a nicer version. This version though is even harder to grasp :D

File: SU2_CFD/src/output/CFlowOutput.cpp (hunk 1 above)
Comment (TobiKattmann) [reply SU2_CFD/src/output/CFlowOutput.cpp:1203]: Ok I guess the benefit is that one won't pollute the screen-output with all surfaces when one only wants the accumulated `TOTAL_HEATFLUX`, so I shall allow it :crown: :D ... making it consistent across the fields stands against consistency in time across versions. Which consistency will prevail? Time will tell...

**Review by TobiKattmann — APPROVED** [review]
Thanks @pcarruscag for tackling that issue, I saw it as well but pretended to haven't seen it :D 

LGTM

My comments are minor things. The only real thing is that I think AVG_TEMPERATURE should be removed for the flow_solver

File: SU2_CFD/src/output/CMultizoneOutput.cpp (hunk 4 above)
Comment (TobiKattmann) [reply SU2_CFD/src/output/CMultizoneOutput.cpp:109]: Tobi: "implements"
Pedro: Lets clean that shit up
:smiling_face_with_tear:

File: SU2_CFD/src/output/CFlowIncOutput.cpp [hunk 5]
```diff
@@ -199,14 +180,17 @@ void CFlowIncOutput::SetHistoryOutputFields(CConfig *config){
...
-
   AddAnalyzeSurfaceOutput(config);
 
-  /*--- Add aerodynamic coefficients fields --- */
-
   AddAerodynamicCoefficients(config);
 
+  if (heat || weakly_coupled_heat) {
+    AddHeatCoefficients(config);
+    AddHistoryOutput("AVG_TEMPERATURE", "Temp", ScreenOutputFormat::SCIENTIFIC, "HEAT", "Average temperature on all surfaces set with MARKER_MONITORING.", HistoryFieldType::COEFFICIENT);
```
[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/src/output/CFlowIncOutput.cpp:189]: What about `weakly_coupled_heat`?

File: SU2_CFD/src/output/CFlowOutput.cpp (hunk 1 above)
Comment (pcarruscag) [reply SU2_CFD/src/output/CFlowOutput.cpp:1203]: Things in marker monitoring use _ON_SURFACE:
`AddHistoryOutputPerSurface("DRAG_ON_SURFACE",       "CD",   ScreenOutputFormat::FIXED, "AERO_COEFF_SURF", Marker_`

File: TestCases/navierstokes/flatplate/lam_flatplate.cfg [hunk 6]
```diff
@@ -11,223 +11,85 @@
...
 % ------------- COMMON PARAMETERS DEFINING THE NUMERICAL METHOD ---------------%
 %
-% Numerical method for spatial gradients (GREEN_GAUSS, WEIGHTED_LEAST_SQUARES)
 NUM_METHOD_GRAD= WEIGHTED_LEAST_SQUARES
-%
-% Courant-Friedrichs-Lewy condition of the finest grid
-CFL_NUMBER= 6.0
-%
-% Adaptive CFL number (NO, YES)
+CFL_NUMBER= 10000
```
[the remark this answers is not in the document]
Comment (pcarruscag) [reply TestCases/navierstokes/flatplate/lam_flatplate.cfg:49]: No it was just horrendously bad

File: SU2_CFD/src/output/CFlowIncOutput.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (TobiKattmann) [reply SU2_CFD/src/output/CFlowIncOutput.cpp:189]: Yep, you're right. There one would use `AVG_TEMPERATURE`

File: SU2_CFD/src/output/CMultizoneOutput.cpp (hunk 4 above)
Comment (pcarruscag) [reply SU2_CFD/src/output/CMultizoneOutput.cpp:109]: Why is it hard to grasp? I also missed it in your PR but it's just a vector.

File: SU2_CFD/src/output/CFlowOutput.cpp (hunk 1 above)
Comment (TobiKattmann) [reply SU2_CFD/src/output/CFlowOutput.cpp:1203]: Uf yes ... i should do my research first. So FLOW_COEFF_SURF is now the only baddie

File: TestCases/navierstokes/flatplate/lam_flatplate.cfg [hunk 7]
```diff
@@ -11,223 +11,85 @@
...
-% 2nd and 4th order artificial dissipation coefficients
-JST_SENSOR_COEFF= ( 0.5, 0.02 )
-%
-% Time discretization (RUNGE-KUTTA_EXPLICIT, EULER_IMPLICIT, EULER_EXPLICIT)
 TIME_DISCRE_FLOW= EULER_IMPLICIT
 
 % --------------------------- CONVERGENCE PARAMETERS --------------------------%
 %
-% Min value of the residual (log10 of the residual)
 CONV_RESIDUAL_MINVAL= -12
```
Comment (pcarruscag) [inline TestCases/navierstokes/flatplate/lam_flatplate.cfg:79]:
```suggestion
CONV_FIELD= RMS_DENSITY
CONV_RESIDUAL_MINVAL= -12
```

File: SU2_CFD/src/output/CMultizoneOutput.cpp (hunk 4 above)
Comment (TobiKattmann) [reply SU2_CFD/src/output/CMultizoneOutput.cpp:109]: That one thing alone of course not, but one needs to dig into how the HistoryOutput works in general and then for multizone things already get quite a bit less obvious than they are for single zone and  PerSurface packs another level of complexity on top.

File: SU2_CFD/src/output/CFlowIncOutput.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/src/output/CFlowIncOutput.cpp:189]: :facepalm: I won't stick my hand in that xD

File: SU2_CFD/src/output/CMultizoneOutput.cpp (hunk 4 above)
Comment (pcarruscag) [reply SU2_CFD/src/output/CMultizoneOutput.cpp:109]: That's true... there is a fair bit of complexity in COutput to allow adding fields easily, still, much much better than legacy.

**Pull request merged**